hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/ClusterBatchedJoinTest.java

```diff
@@ -0,0 +1,116 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.cluster.impl;
+
+import com.hazelcast.cluster.Address;
+import com.hazelcast.config.Config;
+import com.hazelcast.config.JoinConfig;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.internal.util.FutureUtil;
+import com.hazelcast.spi.impl.eventservice.EventRegistration;
+import com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import static com.hazelcast.test.Accessors.getNodeEngineImpl;
+import static org.junit.Assert.assertEquals;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category({QuickTest.class})
+public class ClusterBatchedJoinTest extends HazelcastTestSupport {
+
+    private final HazelcastInstance[] batchedMembers = new HazelcastInstance[10];
+
+    @Before
+    public void prepare() throws InterruptedException {
+        // Prepare members list and config
+        List<String> membersList = new ArrayList<>(10);
+        for (int k = 0; k < 10; k++) {
+            membersList.add("127.0.0.1:" + (5701 + k));
+        }
+        Config config = smallInstanceConfigWithoutJetAndMetrics().setClusterName("myCluster");
+        config.getNetworkConfig().setPort(5701);
+        JoinConfig joinConfig = config.getNetworkConfig().getJoin();
+        joinConfig.getTcpIpConfig().setEnabled(true).setMembers(membersList);
+
+        // Create a cluster that goes through a batched join, starting with a master instance (always index 0)
+        batchedMembers[0] = Hazelcast.newHazelcastInstance(config);
+        // Start 9 members simultaneously (so their join request is batched)
+        List<Future<?>> futures = new ArrayList<>(10);
+        for (int k = 1; k < 10; k++) {
+            int finalK = k;
+            futures.add(spawn(() -> batchedMembers[finalK] = Hazelcast.newHazelcastInstance(config)));
+        }
+
+        // Wait for all member instances to be created
+        FutureUtil.waitWithDeadline(futures, 30, TimeUnit.SECONDS);
+
+        // Ensure cluster size is 10
+        assertClusterSizeEventually(10, batchedMembers);
+    }
+
+    @After
+    public void cleanUp() {
+        // Terminate cluster members
+        for (HazelcastInstance member : batchedMembers) {
+            member.getLifecycleService().terminate();
+        }
+    }
```

Comment:
If you make use of rule: 
```
@Rule
 public final OverridePropertyRule overridePropertyRule = set(HAZELCAST_TEST_USE_NETWORK, "true");
```

You can create instances with test-factory and can get rid of this `cleanUp` method.

Reply:
Even with this override property rule provided, the test will pass regardless of `ClusterJoinManager` state (broken or fixed) - that's why I opted to use a real instance as opposed to test instances

Reply:
maybe i missing something here but with this rule, you will be using real instances.

Reply:
I'm not familiar with the depths of the testing framework, but I did try to get it working, but whatever I tried (following examples from elsewhere in the codebase), my test would always pass using test factories (even with a broken `ClusterJoinManager` implementation).

Reply:
Just note that, I see this a very minor improvement for the test, so not doing it is also ok, but to tell better what i meant, again here a coded version :

```
@Rule
    public final OverridePropertyRule overridePropertyRule = set(HAZELCAST_TEST_USE_NETWORK, "true");

    @Before
    public void prepare() throws InterruptedException {
        // Prepare members list and config
        List<String> membersList = new ArrayList<>(10);
        for (int k = 0; k < 10; k++) {
            membersList.add("127.0.0.1:" + (5701 + k));
        }
        Config config = smallInstanceConfigWithoutJetAndMetrics().setClusterName("myCluster");
        config.getNetworkConfig().setPort(5701);
        JoinConfig joinConfig = config.getNetworkConfig().getJoin();
        joinConfig.getTcpIpConfig().setEnabled(true).setMembers(membersList);

        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory();

        // Create a cluster that goes through a batched join, starting with a master instance (always index 0)
        batchedMembers[0] = factory.newHazelcastInstance(config);
        // Start 9 members simultaneously (so their join request is batched)
        List<Future<?>> futures = new ArrayList<>(10);
        for (int k = 1; k < 10; k++) {
            int finalK = k;
            futures.add(spawn(() -> batchedMembers[finalK] = factory.newHazelcastInstance(config)));
        }

        // Wait for all member instances to be created
        FutureUtil.waitWithDeadline(futures, 30, TimeUnit.SECONDS);

        // Ensure cluster size is 10
        assertClusterSizeEventually(10, batchedMembers);
    }
```

Reply:
Thanks for the example again Ahmet - it turns out I was having trouble because I was using `private final TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory()` as a field variable, and so that was my problem all along! Creating the factory within the `prepare()` method causes everything to run smoothly; another nugget of knowledge for me here 👍

Updated to use test factory here: https://github.com/hazelcast/hazelcast/pull/25114/commits/2111c7133c536d66d50c14e750e15e4cbf01c0d7